internal/config/config.go

```diff
@@ -125,32 +130,39 @@ func configV1alpha1ToConfig(c *configv1alpha1.Config) (*Config, error) {
 		})
 	}
 
-	cfg := &Config{
-		ProjectRoot:             project.GetRoot(),
-		ProjectFilename:         project.GetFilename(),
-		ProjectFindRepoUpward:   project.GetFindRepoUpward(),
-		ProjectIgnorePaths:      project.GetIgnorePaths(),
-		ProjectDisableGitignore: project.GetDisableGitignore(),
-		ProjectEnvUseSystemEnv:  project.GetEnv().GetUseSystemEnv(),
-		ProjectEnvSources:       project.GetEnv().GetSources(),
-		ProjectFilters:          filters,
-
-		RuntimeDockerEnabled:         runtime.GetDocker().GetEnabled(),
-		RuntimeDockerImage:           runtime.GetDocker().GetImage(),
-		RuntimeDockerBuildContext:    runtime.GetDocker().GetBuild().GetContext(),
-		RuntimeDockerBuildDockerfile: runtime.GetDocker().GetBuild().GetDockerfile(),
-
-		ServerAddress:     server.GetAddress(),
-		ServerTLSEnabled:  server.GetTls().GetEnabled(),
-		ServerTLSCertFile: server.GetTls().GetCertFile(),
-		ServerTLSKeyFile:  server.GetTls().GetKeyFile(),
-
-		LogEnabled: log.GetEnabled(),
-		LogPath:    log.GetPath(),
-		LogVerbose: log.GetVerbose(),
+	cfg := Defaults()
+	cfg.ProjectRoot = project.GetRoot()
+	cfg.ProjectFilename = project.GetFilename()
+	cfg.ProjectFindRepoUpward = unwrapBool(project.GetFindRepoUpward(), cfg.ProjectFindRepoUpward)
+	cfg.ProjectIgnorePaths = project.GetIgnorePaths()
+	cfg.ProjectDisableGitignore = unwrapBool(project.GetDisableGitignore(), cfg.ProjectDisableGitignore)
+	cfg.ProjectEnvUseSystemEnv = unwrapBool(project.GetEnv().GetUseSystemEnv(), cfg.ProjectEnvUseSystemEnv)
+	cfg.ProjectEnvSources = project.GetEnv().GetSources()
+	cfg.ProjectFilters = filters
+
+	cfg.RuntimeDockerEnabled = unwrapBool(runtime.GetDocker().GetEnabled(), cfg.RuntimeDockerEnabled)
+	cfg.RuntimeDockerImage = runtime.GetDocker().GetImage()
+	cfg.RuntimeDockerBuildContext = runtime.GetDocker().GetBuild().GetContext()
+	cfg.RuntimeDockerBuildDockerfile = runtime.GetDocker().GetBuild().GetDockerfile()
+
+	cfg.ServerAddress = server.GetAddress()
+	cfg.ServerTLSEnabled = unwrapBool(server.GetTls().GetEnabled(), cfg.ServerTLSEnabled)
+	cfg.ServerTLSCertFile = server.GetTls().GetCertFile()
+	cfg.ServerTLSKeyFile = server.GetTls().GetKeyFile()
+
+	cfg.LogEnabled = unwrapBool(log.GetEnabled(), cfg.LogEnabled)
+	cfg.LogPath = log.GetPath()
+	cfg.LogVerbose = unwrapBool(log.GetVerbose(), cfg.LogVerbose)
+
+	return &cfg, nil
+}
+
+func unwrapBool(wrappedVal *wrapperspb.BoolValue, defaultVal bool) bool {
```

Comment:
Nit: I have a generic alternative:

```go
func setIfHasValue[T any](prop *T, val interface{ GetValue() T }) {
	if val != nil {
		*prop = val.GetValue()
	}
}
```

And usage is as follows:
```go
setIfHasValue(&cfg.ServerTLSEnabled, server.GetTls().GetEnabled())
```

Reply:
love it!